flow-maven-plugin: Problem bundling dependency with multiple bower.json

- Description of the bug

When we try to use a web-component that has a bower depency to popper.js (ex: https://github.com/bisam-rd/genesis-popover), we are not able to package the project in production mode.

- Minimal reproducible example

Use the webjar `org.webjars.bowergithub.fezvrasta:popper.js:1.14.3`.
With this code:

```java

@JavaScript("frontend://bower_components/popper/dist/umd/popper.min.js")
public class GenesisPopover extends Component  {
   ...
}
```

The build with the `flow-maven-plugin` fails.

```
[Exec Stream Pumper] [INFO] Starting to process frontend files.
[Exec Stream Pumper] [INFO] Will minify frontend files.
[Exec Stream Pumper] [INFO] Will bundle frontend files.
[Exec Stream Pumper] [INFO] Will hash bundle file names.
[Exec Stream Pumper] [INFO] Will copy files to target directory 'D:\Dev\env\b-one-web\web-composite\target\build\frontend-es6'.
[Exec Stream Pumper] [INFO] Starting operations stated above, this might take a while.
[Exec Stream Pumper] [ERROR] Failed to process frontend files.
[Exec Stream Pumper] [ERROR] error: Error: ENOENT: no such file or directory, open 'D:\Dev\env\b-one-web\web-composite\target\frontend\bower_components\popper.js\dist\umd\popper.min.js'
```

After further investigation, we found that the bower version of popperjs contains two bower.json files in different directories.

Therefore, the method [selectCorrectPackage](https://github.com/vaadin/flow/blob/40a89b9998c35073db06f4a1073bb56b84835027/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/production/WebJarPackage.java#L89) in the maven-flow-plugin randomly selects one of the directories for packaging.

- Expected behavior

The top most bower.json is used when two bower.json are found in the same webjar ?

- Actual behavior

Randomly selected directory is packaged.

--- a/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/production/WebJarPackage.java
+++ b/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/production/WebJarPackage.java
@@ -16,6 +16,7 @@
 package com.vaadin.flow.plugin.production;
 
 import java.util.Objects;
+import java.util.Optional;
 
 import com.vaadin.flow.plugin.common.ArtifactData;
 
@@ -86,33 +87,46 @@ public class WebJarPackage {
      * @return package with as less issues as possible
      * @throws IllegalArgumentException when packages have different names or versions
      */
-    public static WebJarPackage selectCorrectPackage(WebJarPackage package1, WebJarPackage package2) {
+    public static WebJarPackage selectCorrectPackage(WebJarPackage package1,
+            WebJarPackage package2) {
         if (!Objects.equals(package1.packageName, package2.packageName)) {
             throw new IllegalArgumentException(String.format(
-                    "Cannot process packages with different names: '%s' and '%s'", package1.packageName, package2.packageName));
+                    "Cannot process packages with different names: '%s' and '%s'",
+                    package1.packageName, package2.packageName));
         }
 
-        String normalizedVersion1 = normalizeVersion(package1.webJar.getVersion());
-        String normalizedVersion2 = normalizeVersion(package2.webJar.getVersion());
+        String normalizedVersion1 = normalizeVersion(
+                package1.webJar.getVersion());
+        String normalizedVersion2 = normalizeVersion(
+                package2.webJar.getVersion());
         if (Objects.equals(normalizedVersion1, normalizedVersion2)) {
-            return selectAppropriatePackage(package1, package2, normalizedVersion1);
+            return selectTopmostPackage(package1, package2).orElseGet(
+                    () -> tryToSelectPackageWithNormalizedVersion(package1,
+                            package2, normalizedVersion1));
         }
         throw new IllegalArgumentException(String.format(
-                "Two webJars have same name and different versions: '%s' and '%s', there should be no version differences", package1.webJar, package2.webJar));
+                "Two webJars have same name and different versions: '%s' and '%s', there should be no version differences",
+                package1.webJar, package2.webJar));
     }
-
-    private static WebJarPackage selectAppropriatePackage(
-            WebJarPackage package1, WebJarPackage package2,
-            String normalizedVersion1) {
+    
+    private static Optional<WebJarPackage> selectTopmostPackage(
+            WebJarPackage package1, WebJarPackage package2) {
         String path1 = package1.getPathToPackage();
         String path2 = package2.getPathToPackage();
+
         if (!Objects.equals(path1, path2)) {
             if (path1.startsWith(path2)) {
-                return package2;
+                return Optional.of(package2);
             } else if (path2.startsWith(path1)) {
-                return package1;
+                return Optional.of(package1);
             }
         }
+        return Optional.empty();
+    }
+
+    private static WebJarPackage tryToSelectPackageWithNormalizedVersion(
+            WebJarPackage package1, WebJarPackage package2,
+            String normalizedVersion1) {
         return Objects.equals(normalizedVersion1, package1.webJar.getVersion())
                 ? package1
                 : package2;

--- a/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/production/WebJarPackage.java
+++ b/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/production/WebJarPackage.java
@@ -84,7 +84,7 @@ public class WebJarPackage {
      * @param package1 first package data
      * @param package2 second package data
      * @return package with as less issues as possible
-     * @throws IllegalArgumentException when packages have different names of versions
+     * @throws IllegalArgumentException when packages have different names or versions
      */
     public static WebJarPackage selectCorrectPackage(WebJarPackage package1, WebJarPackage package2) {
         if (!Objects.equals(package1.packageName, package2.packageName)) {
@@ -95,10 +95,27 @@ public class WebJarPackage {
         String normalizedVersion1 = normalizeVersion(package1.webJar.getVersion());
         String normalizedVersion2 = normalizeVersion(package2.webJar.getVersion());
         if (Objects.equals(normalizedVersion1, normalizedVersion2)) {
-            return Objects.equals(normalizedVersion1, package1.webJar.getVersion()) ? package1 : package2;
+            return selectAppropriatePackage(package1, package2, normalizedVersion1);
         }
         throw new IllegalArgumentException(String.format(
                 "Two webJars have same name and different versions: '%s' and '%s', there should be no version differences", package1.webJar, package2.webJar));
+    }
+
+    private static WebJarPackage selectAppropriatePackage(
+            WebJarPackage package1, WebJarPackage package2,
+            String normalizedVersion1) {
+        String path1 = package1.getPathToPackage();
+        String path2 = package2.getPathToPackage();
+        if (!Objects.equals(path1, path2)) {
+            if (path1.startsWith(path2)) {
+                return package2;
+            } else if (path2.startsWith(path1)) {
+                return package1;
+            }
+        }
+        return Objects.equals(normalizedVersion1, package1.webJar.getVersion())
+                ? package1
+                : package2;
     }
 
     private static String normalizeVersion(String version) {

--- a/flow-maven-plugin/src/test/java/com/vaadin/flow/plugin/production/WebJarPackageTest.java
+++ b/flow-maven-plugin/src/test/java/com/vaadin/flow/plugin/production/WebJarPackageTest.java
@@ -66,17 +66,29 @@ public class WebJarPackageTest {
                 packageWithoutPrefixedVersion,
                 createPackage(packageName, prefixedVersion));
 
-        assertEquals("Expected to have version without prefix after merge",
-                merged.getPackageName(),
-                packageWithoutPrefixedVersion.getPackageName());
-        assertEquals("Got different package name after merge",
-                merged.getPathToPackage(),
-                packageWithoutPrefixedVersion.getPathToPackage());
-        assertEquals("Got different WebJar after merge", merged.getWebJar(),
-                packageWithoutPrefixedVersion.getWebJar());
+        assertPackagesEqual(packageWithoutPrefixedVersion, merged);
+    }
+
+    @Test
+    public void selectCorrectPackage_topmostBowerJsonIsSelected() {
+        String name = "popper.js";
+        String version = "1.14.3";
+        WebJarPackage topLevelPackage = createPackage(name, version,
+                "META-INF/resources/webjars/popper.js/");
+        WebJarPackage nestedPackage = createPackage(name, version,
+                "META-INF/resources/webjars/popper.js/packages/popper/");
+        WebJarPackage selected = WebJarPackage
+                .selectCorrectPackage(topLevelPackage, nestedPackage);
+
+        assertPackagesEqual(topLevelPackage, selected);
     }
 
     private WebJarPackage createPackage(String name, String version) {
+        return createPackage(name, version, "path");
+    }
+
+    private WebJarPackage createPackage(String name, String version,
+            String path) {
         if (webJarFile == null) {
             try {
                 webJarFile = testDirectory.newFile("testWebJarFile");
@@ -87,6 +99,15 @@ public class WebJarPackageTest {
         }
         return new WebJarPackage(
                 new ArtifactData(webJarFile, "artifactId", version), name,
-                "path");
+                path);
+    }
+
+    private void assertPackagesEqual(WebJarPackage expected, WebJarPackage actual) {
+        assertEquals("Expected to have version without prefix after merge",
+                expected.getPackageName(), actual.getPackageName());
+        assertEquals("Got different package name after merge",
+                expected.getPathToPackage(), actual.getPathToPackage());
+        assertEquals("Got different WebJar after merge", expected.getWebJar(),
+                actual.getWebJar());
     }
 }

